analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantArgument.cs

```diff
@@ -34,34 +36,28 @@ public sealed class RedundantArgument : SonarDiagnosticAnalyzer
             context.RegisterNodeAction(
                 c =>
                 {
-                    // Can't use optional arguments in expression trees (CS0584), so skip those
-                    if (c.IsInExpressionTree())
+                    if (!c.IsInExpressionTree() // Can't use optional arguments in expression trees (CS0584), so skip those
+                        && ArgumentList(c) is { } argumentList
+                        && new CSharpMethodParameterLookup(argumentList, c.SemanticModel) is { MethodSymbol: { } } methodParameterLookup)
                     {
-                        return;
-                    }
-
-                    var argumentList = c.Node is InvocationExpressionSyntax invocationExpression
-                        ? invocationExpression.ArgumentList
-                        : ((ImplicitObjectCreationExpressionSyntaxWrapper)c.Node).ArgumentList;
-                    var methodParameterLookup = new CSharpMethodParameterLookup(argumentList, c.SemanticModel);
-
-                    if (methodParameterLookup.MethodSymbol != null)
-                    {
-                        foreach (var argumentMapping in methodParameterLookup.GetAllArgumentParameterMappings().Reverse().Where(x => x.Symbol.HasExplicitDefaultValue))
+                        foreach (var argumentMapping in methodParameterLookup.GetAllArgumentParameterMappings().Reverse().Where(x =>
+                            x.Symbol.HasExplicitDefaultValue
+                            && ArgumentHasDefaultValue(x, c.SemanticModel)))
                         {
-                            var hasDefaultValue = ArgumentHasDefaultValue(argumentMapping, c.SemanticModel);
-                            if (argumentMapping.Node.NameColon == null && !hasDefaultValue)
-                            {
-                                return;
-                            }
```

Comment:
This condition was introduced in #4925 which suppresses the rule for no reason (#4620 gives no explanation why this is a problem). I removed it and the code fixes do the right thing still.

Reply:
I think #4620 was meant to cover:
```cs
public class Test
{
    void WithOptional(bool a = false, bool b = false) {}
    void Something()
    {
        WithOptional(false, false); // Noncompliant
                                                  // Noncompliant@-1
        WithOptional(false, true);  // Noncompliant FP
    }
}
```
In the second invocation, the first parameter was provided to allow the next, non-default-one, to be placed there.
Named arguments should solve the issue indeed, but I'm wondering if the community complained about this rule being too noisy for those cases resulting in the "FP" issue. I'm thinking about a method with a lengthy parameter list:
```cs
void MethodWithManyParam(
            string myAmazingUrl = "www.amazing.com", 
            bool isTrue = true,
            bool isFalse = false,
            int isFortyTwo = 42,
            int isZero = 0,
            string superLongParamName = "hello") { }
        
void MyMethod() 
{
    MethodWithManyParam("www.amazing.com", false, true, 0, 42, "World");
    // Maybe this seems easier and more readable than following the rule suggesting to use named arguments:
    MethodWithManyParam(isTrue: false, isFalse: true, isFortyTwo: 0, isZero: 42, superLongParamName: "World");
}
```

Reply:
In that case, I would mark it as "Won't fix" and move on. I think the author of the issue doesn't like named arguments at all 😜.